assets/js/modules/analytics-4/datastore/custom-dimensions-gathering-data.js

```diff
@@ -0,0 +1,395 @@
+					const dataAvailabityDetermined = () =>
+						registry
+							.select( MODULES_ANALYTICS_4 )
+							.selectCustomDimensionDataAvailability(
+								customDimension
+							) !== undefined;
+
+					if ( dataAvailabityDetermined() ) {
+						return true;
+					}
+
+					return new Promise( ( resolve ) => {
+						const unsubscribe = registry.subscribe( () => {
+							if ( dataAvailabityDetermined() ) {
+								unsubscribe();
+								resolve( true );
+							}
+						} );
+					} );
+				}
+		),
+};
+
+const baseReducer = createReducer( ( state, { type, payload } ) => {
+	switch ( type ) {
+		case RECEIVE_CUSTOM_DIMENSION_GATHERING_DATA: {
+			const { customDimension, gatheringData } = payload;
+
+			state.customDimensionsGatheringData[ customDimension ] =
+				gatheringData;
+
+			break;
+		}
+
+		case RECEIVE_CUSTOM_DIMENSION_DATA_AVAILABLE_ON_LOAD: {
+			const { customDimension, dataAvailableOnLoad } = payload;
+
+			state.customDimensionsDataAvailableOnLoad[ customDimension ] =
+				dataAvailableOnLoad;
+
+			break;
+		}
+
+		default: {
+			break;
+		}
+	}
+} );
+
+const baseResolvers = {
+	*isCustomDimensionGatheringData( customDimension ) {
+		const registry = yield Data.commonActions.getRegistry();
+
+		// If the gatheringData flag is already set, return early.
+		if (
+			registry
+				.select( MODULES_ANALYTICS_4 )
+				.isCustomDimensionGatheringData( customDimension ) !== undefined
+		) {
+			return;
+		}
+
+		const dataAvailableOnLoad = registry
+			.select( MODULES_ANALYTICS_4 )
+			.isCustomDimensionDataAvailableOnLoad( customDimension );
+
+		// If dataAvailableOnLoad is true, set gatheringData to false and do nothing else.
+		if ( dataAvailableOnLoad ) {
+			yield baseActions.receiveIsCustomDimensionGatheringData(
+				customDimension,
+				false
+			);
+			return;
+		}
+
+		yield {
+			payload: {
+				customDimension,
+			},
+			type: WAIT_FOR_CUSTOM_DIMENSION_DATA_AVAILABILITY_STATE,
+		};
+
+		const dataAvailability = registry
+			.select( MODULES_ANALYTICS_4 )
+			.selectCustomDimensionDataAvailability( customDimension );
+
+		yield baseActions.receiveIsCustomDimensionGatheringData(
+			customDimension,
+			! dataAvailability
+		);
+
+		if ( dataAvailability ) {
+			yield fetchSaveCustomDimensionDataAvailableStateStore.actions.fetchSaveCustomDimensionDataAvailableState(
+				customDimension
+			);
+		}
+	},
+};
+
+const baseSelectors = {
+	/**
+	 * Determines whether data is available for the custom dimension.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state           Data store's state.
+	 * @param {string} customDimension Custom dimension slug.
+	 * @return {boolean|undefined|null} Returns TRUE if data is available, otherwise FALSE.
+	 *                                  If the request is still being resolved, returns undefined.
+	 *                                  If the data availability can not be determined, returns null.
+	 */
+	selectCustomDimensionDataAvailability: createRegistrySelector(
```

Comment:
This is not a selector, this must be an action. See how we define the `selectProperty` action as an example.

Reply:
Yes, that sounds good to me. Thanks, @techanvil.

Reply:
Ok, I've rewritten as discussed. On reflection it doesn't seem ideal setting `isCustomDimensionGatheringData()` in this action rather than in the `*isCustomDimensionGatheringData()` resolver, but at least it gets rid of the `WAIT_FOR_CUSTOM_DIMENSION_DATA_AVAILABILITY_STATE` control.

We could arguably roll this entire action into the `*isCustomDimensionGatheringData()` resolver, but I think it's probably better to keep it as it is to avoid making the selector a bit of a monolith.